JsonStar.Serialization.Demo/JsonStar.Serialization.Demo3.fst

```diff
@@ -0,0 +1,198 @@
+                        in
+
+                        let prop_json_option_term = call (call (quote JsonStar.Json.getProp) (quote prop_name)) real_arg in 
+
+                        let ctr_callable_term = (mk_function bi (call (pack (Tv_FVar (pack_fv ctr_name))) (pack (Tv_Var b)))) in
+
+                        let match_json = 
+                            `(fun (prop_j:option json) ->
+                                match prop_j with
+                                | Some p -> begin
+                                    match p with
+                                    | JNumber num JInt -> begin 
+                                        let num_v = JsonStar.Parser.int_of_string num in
+                                        let prop_ref : int -> bool = `#prop_refinement_fun in
+                                        if prop_ref num_v 
+                                            then Valid ((`#ctr_callable_term) num_v)
+                                            else Error "BBB"
+                                        end
+                                    | JString s -> Error (P.sprintf "Value string: %s" s)
+                                    | JBoolean b -> Error (P.sprintf "Value bool: %b" b)
+                                    | _ -> Error "Unsupported value"
+                                    end
+                                | _ -> Error (P.sprintf "Unsupported value type")
+                            )
+                        in
+                        let body = call match_json prop_json_option_term in
+                        let final = mk_function json_arg_binder body in
+
+                        final
+                       end
+                    | _ -> fail "Not implemented yet"
+                    end
+                else fail (P.sprintf "Expected record type '%s' to have a single constructor." (implode_qn qname))
+            end
+        | _ -> fail (P.sprintf "Expected record, got %s" (term_to_string (pack (Tv_FVar (pack_fv qname)))))
+        end
+    | None -> fail (P.sprintf "Could not find type %s" (implode_qn qname))
+
+/////////// Function generation part end
+
+// This has to produce a function: (json -> result t)
+// Note the usage of 'fail' for typechecking time errors (when we're dealing with bad usage or unsupported kinds of typ_term as the 
+// structure of the type to deserialize is known during typechecking)
+[@@ noextract_to "FSharp"]
+let deserialize_impl (t : Type) (typ_term : term) : Tac unit = 
+    let e = cur_env() in
+    match inspect (cur_goal ()) with
+    | Tv_Arrow b c -> begin
+        if is_record_type e typ_term
+            then begin
+                let deserializer_f = create_deserializer_f e t b in
+                exact deserializer_f
+                end
+            else fail "We only support deserialization of a record, sorry!"
+        end
+    | _ -> fail "Invalid usage of deserialize_impl_dyn. It should be used to generate a function: 'json -> result t'"
```

Comment:
Error message should refer to `deserialize_impl` not `deserialize_impl_dyn` (I think).

Reply:
Yeah, a leftover from renaming things over time :)